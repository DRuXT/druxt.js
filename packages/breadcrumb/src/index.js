import { DruxtBreadcrumbModule } from './module'

/**
 * The Nuxt.js module function.
 *
 * Installs the module functionality in a Nuxt.js frontend.
 *
 * @type {Function}
 * @exports default
 * @name DruxtBreadcrumbModule
 * @see {@link ./module|DruxtBreadcrumbModule}
 */
export default DruxtBreadcrumbModule

/**
 * Vue.js Mixin.
 *
 * @type {object}
 * @exports DruxtBreadcrumbMixin
 * @see {@link ./mixins/breadcrumb|DruxtBreadcrumbMixin}
 */
 export { DruxtBreadcrumbMixin } from './mixins/breadcrumb'
