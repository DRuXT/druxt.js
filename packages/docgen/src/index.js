import consola from 'consola'
import dmd from 'dmd'
import fs from 'fs'
import globby from 'globby'
import jsdoc2md from 'jsdoc-to-markdown'
import mkdirp from 'mkdirp'
import ncp from 'ncp'
import path from 'path'
import vueDocs from 'vue-docgen-api'

const cwd = path.join(__dirname, '..')

/**
 * DruxtDocgen class.
 */
class DruxtDocgen {
  constructor() {
    this.components = []
    this.destination = 'docs/content'
  }

  /**
   * Generate documentation for source JS and Vue files.
   */
  async generateDocs() {
    consola.info('Generating docs')

    /**
     * Generate API documents.
     */
    await this.generateApiDocs()

    /**
     * Generate list of packages.
     */
    await this.generatePackageList()

    /**
     * Generate list of components.
     */
    await this.generateComponentsList()

    /**
     * Copy external files to the generated files.
     */
    await this.copyFiles()
  }

  /**
   * Copy external files to the generated files.
   */
  async copyFiles() {
    consola.info('Copying files')

    // Changelogs.
    const changelogs = await globby('packages/*/CHANGELOG.md')
    for (const file of changelogs) {
      const module = file.split('/')[1]
      const destination = `${this.destination}/api/packages/${module}/CHANGELOG.md`
      const options = {
        transform: (read, write) => {
          // Append frontmatter
          write.write("---\ntitle: Release notes\n---\n\n")
          read.pipe(write)
        }
      }
      await ncp(file, destination, options, () => {})
    }

    // Contributing guide.
    await ncp('CONTRIBUTING.md', `${this.destination}/guide/CONTRIBUTING.md`)
  }

  /**
   * Generate API documents.
   */
  async generateApiDocs() {
    consola.info('Generating API documents')

    const files = await globby([
      // Javascript.
      'packages/*/src/**/*.js',
      // Vue.js.
      'packages/*/src/components/**/*.vue',
      // Exclude fixtures, mocks and tests.
      '!**/__*__/**/*',
    ])

    for (const file of files) {
      // Get JSDoc template data from file.
      const templateData = jsdoc2md.getTemplateDataSync({
        configure: path.resolve(__dirname, '../jsdoc.json'),
        files: file,
      })

      // Process file based on extension.
      if (file.match(/\.js$/)) {
        await this.processJs(file, templateData)
      }
      else {
        this.components.push({ file, templateData })
        await this.processVue(file, templateData)
      }

      this.writeTemplateData(file, templateData)
    }
  }

  /**
   * Generate list of components.
   */
  async generateComponentsList() {
    consola.info('Generating components list')

    const components = this.components
      .filter((o) => o && o.file.split('/').length === 5 && o.templateData[0])
      .map(({ file, templateData }) =>
        `## ${templateData[0].name}

${templateData[0].description}

[Read the docs](/api/${file.replace('src/', '').replace('.vue', '')})
`)

    const destination = this.destination + '/api/components.md'
    mkdirp.sync(path.dirname(destination))

    fs.writeFileSync(destination, `---
title: Components
---
${components.join('\n* * *\n\n')}

* * *

_Note: The contents of this file where automatically generated by the [Druxt Docgen](https://github.com/druxt/druxt.js/tree/develop/packages/docgen)._
`)
  }

  /**
   * Generate list of packages.
   */
  async generatePackageList() {
    consola.info('Generating package list')

    const packageFiles = await globby('packages/*/package.json')
    const packages = packageFiles
      // Load package data.
      .map((file) => ({
        ...require(path.resolve(__dirname, '../../..', file)),
        __dir: file.split('/')[1],
      }))
      // Filter and sort data
      .filter((o) => !o.private)
      .sort((a, b) => a.name < b.name ? -1 : 1)
      // Process package data.
      .map((pkg) =>
        `## ${pkg.name}

<div class="text-sm">

Version: [${pkg.version}](/api/packages/${pkg.__dir}/CHANGELOG)

</div>

${pkg.description}

[Read the docs](/api/packages/${pkg.__dir}/)
`)

    const destination = this.destination + '/api/README.md'
    mkdirp.sync(path.dirname(destination))

    fs.writeFileSync(destination, `---
title: Packages
---

${packages.join('\n* * *\n\n')}

* * *

_Note: The contents of this file where automatically generated by the [Druxt Docgen](https://github.com/druxt/druxt.js/tree/develop/packages/docgen)._
`)
  }

  /**
   * Process Javascript files.
   */
  async processJs(file, templateData) {
    templateData.map(item => {
      // Vuex state scope fix.
      if (item.name === 'state' && item.scope === 'inner') {
        delete item.scope
      }

      // Vuex mutations/@mutator tag suport.
      if (item.mutators) {
        item.description = item.mutators[0].description
        item.kind = 'method'
        item.scope = 'mutation'
      }
    })
  }

  /**
   * Process Vue.js files.
   */
  async processVue(file, templateData) {
    // Get data from vue-docgen-api.
    const data = await vueDocs.parse(file)

    templateData.map(item => {
      const parts = item.id.split('.')
      if (item.memberof === null && parts.length > 1) {
        parts.pop()
        item.memberof = parts.join('.')
      }

      // @vue-computed tag support.
      if (item._vueComputed) {
        // Strip out HTML table from description.
        item.description = item.description.replace(/<\/p.*/ms, '')

        item._vueComputed.map(property => {
          templateData.push({
            id: `module:${data.displayName}.computed.${property.name}`,
            longname: `module:${data.displayName}.computed.${property.name}`,
            kind: 'property',
            scope: 'static',
            memberof: `module:${data.displayName}.computed`,

            ...property
          })
        })
      }

      // Vue.js Computed properties fixes.
      if (parts[1] === 'computed' && parts.length === 2) {
        item.kind = 'property'
      }
    })
  }

  /**
   * Write template data to file system.
   *
   * @param {string} file - The file name.
   * @param {*} data - The JSDoc / Vuedoc generated data.
   */
  writeTemplateData(file, templateData) {
    if (!templateData) return

    const content = dmd(templateData, {
      'example-lang': 'vue',
      'heading-depth': 2,
      partial: [
        path.resolve(cwd, 'partials/main.hbs'),
        path.resolve(cwd, 'partials/sig-name.hbs')
      ],
      separators: true
    })
    if (!content) return

    const destination = `${this.destination}/api/${file.replace('src/', '').replace(/\.[^/.]+$/, '.md')}`
    mkdirp.sync(path.dirname(destination))

    // Build frontmatter.
    const name = templateData[0].id.split(':').pop()
    const frontmatter = `---\ntitle: ${name}\n---\n\n`

    fs.writeFileSync(destination, frontmatter + content)
  }
}

export { DruxtDocgen }
