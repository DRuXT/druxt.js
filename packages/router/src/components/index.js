/**
 * Vue.js component.
 *
 * @type {object}
 * @exports DruxtRouterComponent
 * @name DruxtRouterComponent
 * @see {@link ./components/DruxtRouter.html|DruxtRouterComponent}
 *
 * @example
 * <script>
 * import { DruxtRouterComponent } from 'druxt-router/dist/components'
 *
 * export default {
 *   components: { DruxtRouterComponent }
 * }
 * </script>
 */
export { default as DruxtRouterComponent } from './DruxtRouter.vue'
