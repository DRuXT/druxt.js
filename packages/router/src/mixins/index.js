/**
 * Router Entity Vue.js Mixin.
 *
 * @type {object}
 * @exports DruxtRouterEntityMixin
 * @name DruxtRouterEntityMixin
 * @see {@link ./mixins/entity|DruxtRouterEntityMixin}
 *
 * @example
 * <script>
 * import { DruxtRouterEntityMixin } from 'druxt-router'
 *
 * export default {
 *   mixins: [DruxtRouterEntityMixin]
 * }
 * </script>
 */
 export { DruxtRouterEntityMixin } from './entity'

/**
 * Router Wrapper Vue.js Mixin.
 *
 * @type {object}
 * @exports DruxtRouterMixin
 * @name DruxtRouterMixin
 * @see {@link ./mixins/router|DruxtRouterMixin}
 *
 * @example
 * <script>
 * import { DruxtRouterMixin } from 'druxt-router'
 *
 * export default {
 *   mixins: [DruxtRouterMixin]
 * }
 * </script>
 */
export { DruxtRouterMixin } from './router'
