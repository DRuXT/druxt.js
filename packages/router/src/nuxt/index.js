import { existsSync } from 'fs'
import { join, resolve } from 'path'
import { DrupalJsonApiParams } from 'drupal-jsonapi-params'
import { DruxtClient } from 'druxt'

/**
 * The Nuxt.js module function.
 *
 * - Extends the Vue router, adding the Druxt wildcard route.
 * - Adds the Druxt router plugin.
 * - Adds the Druxt router Vuex store.
 *
 * The module function should not be used directly, but rather installed via your Nuxt configuration file.
 *
 * Options are set on the root level `druxt` Nuxt config object.
 *
 * @example @lang js
 * // `nuxt.config.js`
 * module.exports = {
 *   buildModules: ['druxt-router/nuxt'],
 *   druxt: {
 *     baseUrl: 'https://demi-api.druxtjs.org'
 *   }
 * }
 *
 * @property {object} options.druxt - Druxt root level options.
 * @property {string} options.druxt.baseUrl - Base URL of Drupal JSON:API backend.
 * @property {string} options.druxt.router.component - File to custom Router component.
 */
const DruxtRouterNuxtModule = async function (moduleOptions = {}) {
  // Set default options.
  const options = {
    baseUrl: moduleOptions.baseUrl,
    ...(this.options || {}).druxt || {},
    router: {
      pages: (await existsSync(resolve(this.options.srcDir, this.options.dir.pages))),
      wildcard: true,
      ...((this.options || {}).druxt || {}).router,
      ...moduleOptions,
    }
  }

  // Register components directories.
  this.nuxt.hook('components:dirs', dirs => {
    dirs.push({ path: join(__dirname, '../dist/components') })
  })

  // Add Druxt router custom wildcard route.
  if (options.router.wildcard) {
    // Ignore page routes.
    if (!options.router.pages) {
      this.nuxt.hook('build:before', () =>
        this.nuxt.options.build.createRoutes = () => []
      )
    }

    // Add route template.
    this.addTemplate({
      src: resolve(__dirname, '../templates/component.js'),
      fileName: 'components/druxt-router.js',
      options
    })

    // Fetch languages.
    let languages = []
    const druxt = new DruxtClient(options.baseUrl, {
      ...options,
      // Disable API Proxy, as Proxies aren't available at build.
      proxy: { ...options.proxy || {}, api: false },
    })
    const languageResourceType = 'configurable_language--configurable_language'
    if (await druxt.getIndex(languageResourceType)) {
      const query = new DrupalJsonApiParams().addFields(languageResourceType, ['drupal_internal__id'])
      languages = (await druxt.getCollectionAll(languageResourceType, query) || [])
        .map((o) => o.data)
        .flat()
        .filter((o) => !['und', 'zxx'].includes(((o || {}).attributes || {}).drupal_internal__id))
    }

    // Extend routes.
    this.extendRoutes((routes) => {
      // Add route per language.
      languages.forEach((o) => {
        routes.push({
          name: `druxt-router__${o.attributes.drupal_internal__id}`,
          path: `/${o.attributes.drupal_internal__id}*`,
          component: resolve(this.options.buildDir, 'components/druxt-router.js'),
          chunkName: 'druxt-router',
          meta: { langcode: o.attributes.drupal_internal__id }
        })
      })

      // Add wildcard route.
      routes.push({
        name: 'druxt-router',
        path: '*',
        component: resolve(this.options.buildDir, 'components/druxt-router.js'),
        chunkName: 'druxt-router'
      })
    })
  }

  // Add plugin.
  this.addPlugin({
    src: resolve(__dirname, '../templates/plugin.js'),
    fileName: 'druxt-router.js',
    options
  })

  // Enable Vuex Store.
  this.options.store = true

  // Add Vuex plugin.
  this.addPlugin({
    src: resolve(__dirname, '../templates/store.js'),
    fileName: 'store/druxt-router.js',
    options
  })

  // Nuxt Storybook.
  this.nuxt.hook('storybook:config', async ({ stories }) => {
    this.addTemplate({
      src: resolve(__dirname, '../templates/druxt-router.stories.js'),
      fileName: 'stories/druxt-router.stories.js',
      options: {}
    })
    stories.push(resolve(this.options.buildDir, './stories/druxt-router.stories.js'))
  })
}

DruxtRouterNuxtModule.meta = require('../package.json')

export default DruxtRouterNuxtModule
