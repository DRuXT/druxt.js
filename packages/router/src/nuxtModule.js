import { join, resolve } from 'path'

/**
 * The Nuxt.js module function.
 *
 * - Extends the Vue router, adding the Druxt wildcard route.
 * - Adds the Druxt router plugin.
 * - Adds the Druxt router Vuex store.
 *
 * The module function should not be used directly, but rather installed via your Nuxt configuration file.
 *
 * Options are set on the root level `druxt` Nuxt config object.
 *
 * @example @lang js
 * // `nuxt.config.js`
 * module.exports = {
 *   modules: [
 *     'druxt-router'
 *   ],
 *   druxt: {
 *     baseUrl: 'https://demi-api.druxtjs.org'
 *   }
 * }
 *
 * @property {object} options.druxt - Druxt root level options.
 * @property {string} options.druxt.baseUrl - Base URL of Drupal JSON:API backend.
 * @property {string} options.druxt.router.component - File to custom Router component.
 */
const DruxtRouterNuxtModule = function (moduleOptions) {
  // Set default options.
  const options = {
    baseUrl: moduleOptions.baseUrl,
    ...(this.options || {}).druxt || {},
    router: {
      pages: true,
      wildcard: true,
      ...((this.options || {}).druxt || {}).router,
      ...moduleOptions,
    }
  }

  // Register components directories.
  this.nuxt.hook('components:dirs', dirs => {
    dirs.push({ path: join(__dirname, 'components') })
  })

  // Add Druxt router custom wildcard route.
  if (options.router.wildcard) {
    this.addTemplate({
      src: resolve(__dirname, '../nuxt/component.js'),
      fileName: 'components/druxt-router.js',
      options
    })

    this.extendRoutes((routes, resolve) => {
      routes.push({
        name: 'druxt-router',
        path: '*',
        component: resolve(this.options.buildDir, 'components/druxt-router.js'),
        chunkName: 'druxt-router'
      })
    })
  }

  // Ignore page routes.
  if (options.router.wildcard && !options.router.pages) {
    this.nuxt.hook('build:before', () =>
      this.nuxt.options.build.createRoutes = () => []
    )
  }

  // Add plugin.
  this.addPlugin({
    src: resolve(__dirname, '../nuxt/plugin.js'),
    fileName: 'druxt-router.js',
    options
  })

  // Enable Vuex Store.
  this.options.store = true

  // Add Vuex plugin.
  this.addPlugin({
    src: resolve(__dirname, '../nuxt/store.js'),
    fileName: 'store/druxt-router.js',
    options
  })
}

export { DruxtRouterNuxtModule }
