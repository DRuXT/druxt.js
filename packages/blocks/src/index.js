import { DruxtBlocksNuxtModule } from './nuxtModule'

/**
 * The Nuxt.js module function.
 *
 * Installs the module functionality in a Nuxt application.
 *
 * @type {Function}
 * @exports default
 * @name DruxtBlocksModule
 * @see {@link ./nuxtModule|DruxtBlocksModule}
 *
 * @example <caption>nuxt.config.js</caption> @lang js
 * module.exports = {
 *   modules: [
 *     'druxt-blocks'
 *   ],
 *   druxt: {
 *     baseUrl: 'https://demo-api.druxtjs.org'
 *   }
 * }
 */
export default DruxtBlocksNuxtModule

/**
 * Vue.js Mixin.
 *
 * @type {object}
 * @exports DruxtBlocksBlockMixin
 * @see {@link ./mixins/block|DruxtBlocksBlockMixin}
 */
 export { DruxtBlocksBlockMixin } from './mixins/block'
