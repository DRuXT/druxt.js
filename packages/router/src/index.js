import { DruxtRouterNuxtModule } from './nuxtModule'

/**
 * The Nuxt.js module function.
 *
 * Installs the module functionality in a Nuxt application.
 *
 * @type {Function}
 * @exports default
 * @name DruxtRouterNuxtModule
 * @see {@link ./nuxtModule|DruxtRouterNuxtModule}
 *
 * @example <caption>nuxt.config.js</caption> @lang js
 * module.exports = {
 *   modules: [
 *     'druxt-router'
 *   ],
 *   druxt: {
 *     baseUrl: 'https://demo-api.druxtjs.org'
 *   }
 * }
 */
export default DruxtRouterNuxtModule

/**
 * The DruxtRouter class.
 *
 * Core functionality for the DruxtRouter module.
 *
 * @type {class}
 * @exports DruxtRouter
 * @name DruxtRouter
 * @see {@link ./router.html|DruxtRouter}
 */
export { DruxtRouter } from './router'

/**
 * Vuex store module.
 *
 * @type {object}
 * @exports DruxtRouterStore
 * @name DruxtRouterStore
 * @see {@link ./stores/router|DruxtRouterStore}
 */
export { DruxtRouterStore } from './stores/router'

/**
 * Router Entity Vue.js Mixin.
 *
 * @type {object}
 * @exports DruxtRouterEntityMixin
 * @name DruxtRouterEntityMixin
 * @see {@link ./mixins/entity|DruxtRouterEntityMixin}
 *
 * @example
 * <script>
 * import { DruxtRouterEntityMixin } from 'druxt-router'
 *
 * export default {
 *   mixins: [DruxtRouterEntityMixin]
 * }
 * </script>
 */
 export { DruxtRouterEntityMixin } from './mixins/entity'

 /**
  * Router Wrapper Vue.js Mixin.
  *
  * @type {object}
  * @exports DruxtRouterMixin
  * @name DruxtRouterMixin
  * @see {@link ./mixins/router|DruxtRouterMixin}
  *
  * @example
  * <script>
  * import { DruxtRouterMixin } from 'druxt-router'
  *
  * export default {
  *   mixins: [DruxtRouterMixin]
  * }
  * </script>
  */
 export { DruxtRouterMixin } from './mixins/router'
 