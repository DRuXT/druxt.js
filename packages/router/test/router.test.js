import mockAxios from 'jest-mock-axios'

import { DruxtRouter } from '../src'

const baseURL = 'https://demo-api.druxtjs.org'

const testArticle = { type: 'node--article', id: '98f36405-e1c4-4d8a-a9f9-4d4f6d414e96' }
const testPage = { type: 'node--page', id: '4eb8bcc1-3b2e-4663-89cd-b8ca6d4d0cc9' }

let router

jest.mock('axios')

describe('DruxtRouter', () => {
  beforeEach(() => {
    mockAxios.reset()
    router = new DruxtRouter(baseURL, {})
  })

  test('constructor', () => {
    // Throw error if 'baseURL' not provided.
    expect(() => { return new DruxtRouter() }).toThrow('The \'baseUrl\' parameter is required.')

    // Ensure class type.
    expect(new DruxtRouter(baseURL)).toBeInstanceOf(DruxtRouter)
  })

  // @deprecated
  test('constructor - axiosSettings', () => {
    const headers = { 'X-DruxtRouter': true }
    const mockRouter = new DruxtRouter(baseURL, {
      axios: { headers }
    })
    expect(mockRouter).toBeInstanceOf(DruxtRouter)

    expect(mockAxios.create).toHaveBeenCalledWith({ baseURL, headers })
  })

  // @deprecated
  test('addHeaders', () => {
    expect(router.addHeaders()).toBe(false)

    router.addHeaders({ 'X-DruxtRouter': true })
    expect(router.axios.defaults.headers.common['X-DruxtRouter']).toBe(true)
  })

  // @deprecated
  test('buildQueryUrl', () => {
    expect(router.buildQueryUrl('url', 'query')).toBe('url?query')
    expect(router.buildQueryUrl('url', '?query')).toBe('url?query')
    expect(router.buildQueryUrl('url', { getQueryString: () => 'query' })).toBe('url?query')
    expect(router.buildQueryUrl('url', { query: 'string' })).toBe('url?query=string')

    expect(router.buildQueryUrl('url', {})).toBe('url')
    expect(router.buildQueryUrl('url', [])).toBe('url')
    expect(router.buildQueryUrl('url', Boolean)).toBe('url')
    expect(router.buildQueryUrl('url')).toBe('url')
  })

  // @deprecated
  test('checkPermissions', () => {
    const res = {
      data: {
        meta: {
          omitted: {
            detail: 'Some resources have been omitted because of insufficient authorization.',
            links: {
              help: null,
              item: {
                meta: {
                  detail: 'The current user is not allowed to GET the selected resource. The \'administer node fields\' permission is required.'
                }
              }
            }
          }
        }
      }
    }
    expect(() => router.checkPermissions(res)).toThrow('Some resources have been omitted because of insufficient authorization.\n\n Required permissions: administer node fields.')
  })

  test('get - entity', async () => {
    const { route } = await router.get('/')

    expect(route.component).toBe('druxt-entity')
    expect(route.type).toBe('entity')
    expect(route.props).toHaveProperty('type')
    expect(route.props).toHaveProperty('uuid')
  })

  test('get - views', async () => {
    const { route } = await router.get('/view')

    expect(route.component).toBe('druxt-view')
    expect(route.type).toBe('views')
    expect(route.props).toHaveProperty('displayId')
    expect(route.props).toHaveProperty('type')
    expect(route.props).toHaveProperty('uuid')
    expect(route.props).toHaveProperty('viewId')
  })

  test('get - error', async () => {
    const { route } = await router.get('/error')

    expect(route.error).toHaveProperty('message')
    expect(route.error).toHaveProperty('statusCode', 404)
  })

  // @deprecated
  test('getIndex', async () => {
    const index = await router.getIndex()
    expect(mockAxios.get).toHaveBeenCalledTimes(2)
    expect(mockAxios.get).toHaveBeenCalledWith('/jsonapi')

    expect(Object.keys(index).length).toBe(54)
    expect(index[Object.keys(index)[0]]).toHaveProperty('href')

    const cachedIndex = await router.getIndex()
    expect(mockAxios.get).toHaveBeenCalledTimes(2)

    expect(Object.keys(cachedIndex).length).toBe(54)
  })

  // @deprecated
  test('getIndex - resource', async () => {
    const resourceIndex = await router.getIndex('node--page')
    expect(mockAxios.get).toHaveBeenCalledTimes(2)
    expect(mockAxios.get).toHaveBeenCalledWith('/jsonapi')

    expect(resourceIndex).toHaveProperty('href')

    const cachedResourceIndex = await router.getIndex('node--page')
    expect(mockAxios.get).toHaveBeenCalledTimes(2)

    expect(cachedResourceIndex).toHaveProperty('href')
  })

  test('getRedirect', () => {
    let redirect

    // Route provided redirect.
    redirect = router.getRedirect(null, {
      redirect: [{ to: '/' }]
    })
    expect(redirect).toBe('/')

    // Root redirect.
    redirect = router.getRedirect('/', {
      isHomePath: true
    })
    expect(redirect).toBe(false)

    redirect = router.getRedirect('/node/1', {
      isHomePath: true
    })
    expect(redirect).toBe('/')

    // Clean url redirect.
    redirect = router.getRedirect('/node/2', {
      canonical: 'https://example.com/clean-url'
    })
    expect(redirect).toBe('/clean-url')

    // No redirect.
    redirect = router.getRedirect(null, {})
    expect(redirect).toBe(false)

    // Querystring.
    redirect = router.getRedirect('/?querystring', {
      isHomePath: true
    })
    expect(redirect).toBe(false)

    redirect = router.getRedirect('/clean-url?querystring', {
      canonical: 'https://example.com/clean-url'
    })
    expect(redirect).toBe(false)
  })

  // @deprecated
  test('getResource', async () => {
    const entity = await router.getResource(testArticle)
    expect(entity).toHaveProperty('type', testArticle.type)

    const error = await router.getResource({ id: 'test', type: 'missing' })
    expect(mockAxios.get).toHaveBeenCalledWith('/jsonapi/missing/test')
    expect(error).toBe(false)

    const empty = await router.getResource()
    expect(empty).toBe(false)
  })

  // @deprecated
  test('getResources', async () => {
    const resources = await router.getResources('node--page')
    expect(mockAxios.get).toHaveBeenLastCalledWith('/jsonapi/node/page')
    expect(resources.length).toBe(1)

    await router.getResources('node--page', { 'filter[status]': 1 })
    expect(mockAxios.get).toHaveBeenLastCalledWith('/jsonapi/node/page?filter%5Bstatus%5D=1')

    const noResource = await router.getResources()
    expect(noResource).toBe(false)

    await router.getResources('test--all', null, { all: true })
    expect(mockAxios.get).toHaveBeenLastCalledWith('/jsonapi/test/all?next')
  })

  test('getResourceByRoute', async () => {
    const route = require('../src/__fixtures__/data/0a01adaa07e9dfcc3c0cabc37339505a.json')
    const entity = await router.getResourceByRoute(route)

    expect(entity).toHaveProperty('id', testPage.id)
    expect(entity).toHaveProperty('type', testPage.type)
  })

  test('getRoute', async () => {
    // Get the route of the homepage.
    let route = await router.getRoute('/')

    expect(route).toHaveProperty('canonical')
    expect(route).toHaveProperty('component')
    expect(route).toHaveProperty('error')
    expect(route).toHaveProperty('isHomePath', true)
    expect(route).toHaveProperty('jsonapi')
    expect(route).toHaveProperty('label')
    expect(route).toHaveProperty('props')
    expect(route).toHaveProperty('redirect')
    expect(route).toHaveProperty('type')

    // Get the route of node/1.
    route = await router.getRoute('/node/1')

    expect(route).toHaveProperty('isHomePath', false)
  })
})
