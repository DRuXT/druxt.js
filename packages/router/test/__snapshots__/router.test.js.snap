// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DruxtRouter checkPermissions 1`] = `
{
  "detail": "Required permissions:
 - administer node fields",
}
`;

exports[`DruxtRouter getRoute 1`] = `
{
  "config": {
    "baseURL": "https://8080-magenta-koi-31w8trra.ws-us25.gitpod.io",
    "headers": {
      "Accept": "application/json, text/plain, */*",
    },
    "maxBodyLength": -1,
    "maxContentLength": -1,
    "method": "get",
    "options": {},
    "timeout": 0,
    "transformRequest": [
      null,
    ],
    "transformResponse": [
      null,
    ],
    "url": "/router/translate-path?path=/broken-router",
    "xsrfCookieName": "XSRF-TOKEN",
    "xsrfHeaderName": "X-XSRF-TOKEN",
  },
  "data": {
    "errors": [
      {
        "detail": "Please ensure the Decoupled Router module is installed and configured correctly.",
      },
    ],
  },
  "headers": {
    "cache-control": "must-revalidate, no-cache, private",
    "content-language": "en",
    "content-type": "text/html; charset=UTF-8",
    "expires": "Sun, 19 Nov 1978 05:00:00 GMT",
  },
  "request": {},
  "status": 404,
  "statusText": "Not Found",
}
`;
